platform/saivpp/vpplib/SwitchStateBaseAcl.cpp

```diff
@@ -597,65 +898,100 @@ sai_status_t SwitchStateBase::AclTblConfig(
 				meta->attridname);
 	    }
 	    
-	    if (attr->id == SAI_ACL_ENTRY_ATTR_FIELD_ACL_RANGE_TYPE) {
+        if(attr->id == SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED) {
+            if(attr->value.aclfield.data.booldata == true) {
+                update_tunterm_rule = true;
+            }
+        }
+
+	    if (attr->id == SAI_ACL_ENTRY_ATTR_FIELD_ACL_RANGE_TYPE && acl->count>0  && (rule<(&acl->rules[0]+n_entries))) {
 		for (uint32_t jdx = 0; jdx < p_ace->range_count; jdx++) {
 		    status = acl_rule_port_range_vpp_acl_set(p_ace->range_type[jdx],
 							     &p_ace->range_limit[jdx], rule);
 		}
 	    } else {
-		status = acl_rule_field_update((sai_acl_entry_attr_t) attr->id, &attr->value, rule);
-	    }
+            if(acl->count>0 && (rule<(&acl->rules[0]+n_entries))) {
+		        status = acl_rule_field_update((sai_acl_entry_attr_t) attr->id, &attr->value, rule);
+            }
+            if(tunterm_acl->count>0 && (tunterm_rule<(&tunterm_acl->rules[0]+n_tunterm_entries))) {
+                status = tunterm_acl_rule_field_update((sai_acl_entry_attr_t) attr->id, &attr->value, tunterm_rule);
+	        }
+        }
 	}
-	rule++;
+    
+    if(update_tunterm_rule) {
+        if(acl->count>0 && (rule<(&acl->rules[0]+n_entries))) {
+            memset(rule, 0, sizeof(*rule)); // clear any fields that were added to regular rule
```

Comment:
Can you explain this?

Reply:
When iterating through the incoming rule's sai attributes we don't know if it's a tunterm acl or regular acl until SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is encountered. There are common attr between the two, like SAI_ACL_ENTRY_ATTR_FIELD_DST_IP, which would need to be populated in both cases. Both rules (regular acl and tunterm acl) are updated in parallel while looping through the sai attributes, and then at the end once we know if it's a tunterm rule or not we clear the other rule which potentially had fields populated.